core/src/main/scala/com/databricks/labs/remorph/graph/TableDependencyGraph.scala

```diff
@@ -0,0 +1,139 @@
+    var toTable: TableDefinition = null
+    var action = "SELECT"
+    var fromTable: Seq[TableDefinition] = Seq.empty
+
+    def collectTables(node: ir.LogicalPlan): Unit = {
+      node match {
+        case _: ir.CreateTable =>
+          toTable = tableDefinition.filter(_.table == getTableName(plan)).head
+          action = "CREATE"
+        case _: ir.InsertIntoTable =>
+          toTable = tableDefinition.filter(_.table == getTableName(plan)).head
+          action = "INSERT"
+        case _: ir.DeleteFromTable =>
+          toTable = tableDefinition.filter(_.table == getTableName(plan)).head
+          action = "DELETE"
+        case _: ir.UpdateTable =>
+          toTable = tableDefinition.filter(_.table == getTableName(plan)).head
+          action = "UPDATE"
+        case _: ir.MergeIntoTable =>
+          toTable = tableDefinition.filter(_.table == getTableName(plan)).head
+          action = "MERGE"
+        case _: ir.Project | _: ir.Join | _: ir.SubqueryAlias | _: ir.Filter =>
+          val tableList = plan collect { case x: ir.NamedTable =>
+            x.unparsed_identifier
+          }
+          fromTable = tableDefinition.filter(x => tableList.contains(x.table))
+        case _ => // Do nothing
+      }
+      node.children.foreach(collectTables)
+    }
+
+    collectTables(plan)
+
+    if (fromTable.isEmpty) {
+      logger.warn(s"No tables found for insert into table values query")
+    } else {
+      fromTable.foreach(f => {
+        if (f != toTable) {
+          addEdge(f, toTable, Map("query" -> queryId, "action" -> action))
+        } else {
+          logger.warn(s"Ignoring reference detected for table ${f.table}")
+        }
+      })
+    }
+  }
+
+  private def buildNode(plan: ir.LogicalPlan, tableDefinition: Seq[TableDefinition], query: ExecutedQuery): Unit = {
+    plan collect { case x: ir.NamedTable =>
+      print(x.unparsed_identifier)
+      print("\n")
+      tableDefinition.find(_.table == x.unparsed_identifier) match {
+        case Some(name) => addNode(name, Map("query" -> query.id))
+        case None =>
+          logger.warn(
+            s"Table ${x.unparsed_identifier} not found in table definitions " +
+              s"or is it a subquery alias")
+      }
+    }
+  }
+
+  def buildDependency(queryHistory: QueryHistory, tableDefinition: Seq[TableDefinition]): Unit = {
+    queryHistory.queries.foreach { query =>
+      print("\n")
+      print(query.source)
+      print("\n")
+      val plan = parser.parse(SourceCode(query.source)).flatMap(parser.visit)
+      plan match {
+        case Result.Success(p) => buildNode(p, tableDefinition, query)
+          generateEdges(p, tableDefinition, query.id)
+        case _ => logger.warn(s"Failed to produce plan from query: ${query.source}")
+      }
+    }
+  }
+
+  // TODO Implement logic for fetching edges(parents) only upto certain level
+  override def getRoot(table: String, level: Int = 1): TableDefinition = {
```

Comment:
there are multiple roots. the simplest logic should be:
- count number of indegrees for a node. indegree is an incoming write to a table.
- tables with zero indegrees is a root, as nothing writes to it.

Reply:
Done